src/core/ddsi/src/ddsi__typelib.h

```diff
@@ -66,10 +63,27 @@ dds_return_t ddsi_type_ref_proxy (struct ddsi_domaingv *gv, struct ddsi_type **t
 dds_return_t ddsi_type_add_typeobj (struct ddsi_domaingv *gv, struct ddsi_type *type, const struct DDS_XTypes_TypeObject *type_obj);
 
 /** @component type_system */
-dds_return_t ddsi_type_get_typeinfo_ser (struct ddsi_domaingv *gv, const struct ddsi_type *type, unsigned char **data, uint32_t *sz);
+dds_return_t ddsi_type_get_typeinfo_ser (struct ddsi_domaingv *gv, const struct ddsi_type *type_c, unsigned char **data, uint32_t *sz);
 
-/** @component type_system */
-dds_return_t ddsi_type_get_typeinfo (struct ddsi_domaingv *gv, const struct ddsi_type *type, struct ddsi_typeinfo *type_info);
+
+/**
+ * @brief Gets the type_info for a complete type
+ * @component type_system
+ *
+ * This functions gets the type information structure for the provided
+ * complete type. The provided type should exist in the type library and
+ * should be resolved. The corresponding minimal type is calculated from
+ * the complete type, and if it is not in the type library, it is added
+ * with a refcount of 1. This applies to the top-level minimal type and
+ * all underlying dependencies.
+ *
+ * @param gv  domain globals
+ * @param type_c  the complete DDSI type
+ * @param type_info  the type_info structure to create
+ * @param type_m  the minimal DDSI type corresponding to the provided complete type is stored in this variable
+ * @return dds_return_t
+ */
+dds_return_t ddsi_type_get_typeinfo_locked (struct ddsi_domaingv *gv, const struct ddsi_type *type_c, struct ddsi_typeinfo *type_info, struct ddsi_type **type_m);
```

Comment:
It appears from the rest of the code that in all (successful) cases the `type_c` and `*type_m` point to the complete and corresponding minimal type and have this pointer reflected in their reference counts. (I don't know how to put it otherwise ...)

For the minimal type this is suggested by the description, although it only mentions the case where the type is added. For the complete type, there is no hint whatsoever that this is the case, and it is contradictory to the `const` bit. I am always a bit ambivalent about `const` vs. reference counts and mutexes, whether the object changes from an application perspective or from a memory contents perspective ...

`const` in C really only means the second, and so only says it is ROMmable. In reality, that's often the less useful thing ... All of these `ddsi_type`s are constructed dynamically, so cannot be in ROM, so I am generally not against a refcount changing during the operation. In this case however, it is not merely that the function has a net-zero effect on the refcount.

So I do think it would be better to mark it non-`const`.